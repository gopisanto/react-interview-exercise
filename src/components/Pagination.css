:local(.pagination) {
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 5px 0;
}
:local(.pagination span) {
    cursor: pointer;
    text-decoration: underline;
    color: white;
    border: 1px solid white;
    padding: 5px;
}
:local(span.disabled) {
    color: rgb(15, 15, 15);
    text-decoration: none;
    cursor: default;
    pointer-events: none;
}
:local(span.paginationStatus) {
    margin-left: 5px;
    border: none;
    text-decoration: none;
}