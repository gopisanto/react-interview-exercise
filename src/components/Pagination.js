import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { range } from 'lodash';
import classnames from 'classnames';
import styles from './Pagination.css';

class Pagination extends Component {
    constructor(props) {
        super(props);
        this.state = {
            startPage: 1,
            endPage: props.noOfPageLinksToDisplay - 1,
            activePage: props.defaultActivePage
        }

        this.totalPages = this.totalPages.bind(this);
        this.calculateStartPage = this.calculateStartPage.bind(this);
        this.setSelectedPage = this.setSelectedPage.bind(this);
        this.firstPage = this.firstPage.bind(this);
        this.lastPage = this.lastPage.bind(this);
        this.nextPage = this.nextPage.bind(this);
        this.prevPage = this.prevPage.bind(this);
        this.handlePageClick = this.handlePageClick.bind(this);
        this.generatePageLinks = this.generatePageLinks.bind(this);
    }

    totalPages() {
        return Math.ceil(this.props.noOfItems / this.props.noOfItemsPerPage);
    }

    calculateStartPage(selectedPage) {
        const {startPage} = this.state;
        const {noOfPageLinksToDisplay} = this.props;
        const totalPages = this.totalPages();
        const endPage = Math.min(startPage + noOfPageLinksToDisplay - 1, totalPages);

        if (selectedPage < startPage) {
            return selectedPage === 1 ? selectedPage : selectedPage - 1;
        } else if (selectedPage > endPage) {
            return startPage + selectedPage - endPage; 
        } else if (selectedPage === totalPages) {
            return startPage;
        } else if (startPage + noOfPageLinksToDisplay - 1 === selectedPage && selectedPage < totalPages) {
            return startPage + 1;
        } else if (selectedPage === startPage && startPage !== 1) {
            return startPage - 1;
        }
        return startPage;
    }

    setSelectedPage(selectedPage) {
        this.setState({activePage: selectedPage, startPage: this.calculateStartPage(selectedPage)});
        this.props.onPageChange(selectedPage);
    }

    firstPage() {
        this.setSelectedPage(1);
    }

    lastPage() {
        this.setSelectedPage(this.totalPages());
    }

    nextPage() {
        this.setSelectedPage(this.state.activePage + 1);
    }

    prevPage() {
        this.setSelectedPage(this.state.activePage - 1);
    }

    handlePageClick(selectedPage) {
        return () => {
            this.setSelectedPage(selectedPage);
        }
    }

    generatePageLinks() {
        const {activePage, startPage} = this.state;
        const {noOfPageLinksToDisplay, showStatus, noOfItems} = this.props;
        const totalPages = this.totalPages();
        const endPage = Math.min(startPage + noOfPageLinksToDisplay - 1, totalPages);
        const pageLinks = range(startPage, endPage + 1)
            .map(pageNum => (
                    <span
                        className={classnames({[styles.disabled]: pageNum === activePage})}
                        key={pageNum}
                        onClick={this.handlePageClick(pageNum)}>
                            {pageNum}
                    </span>
                )
            );
        const firstAndPrevLinkDisabled = activePage === 1;
        const lastAndNextLinkDisabled = activePage === totalPages;

        if (totalPages === 1 || noOfItems === 0) {
            return null;
        }
        
        return (
            <div className={styles.pagination}>
                <span
                    className={classnames({[styles.disabled]: firstAndPrevLinkDisabled})}
                    disabled={firstAndPrevLinkDisabled}
                    onClick={this.firstPage}>
                        {'<<'}
                    </span>
                <span
                    className={classnames({[styles.disabled]: firstAndPrevLinkDisabled})}
                    disabled={firstAndPrevLinkDisabled}
                    onClick={this.prevPage}>
                        {'<'}
                    </span>
                {pageLinks}
                <span
                    className={classnames({[styles.disabled]: lastAndNextLinkDisabled})}
                    disabled={lastAndNextLinkDisabled}
                    onClick={this.nextPage}>
                        {'>'}
                    </span>
                <span
                    className={classnames({[styles.disabled]: lastAndNextLinkDisabled})}
                    disabled={lastAndNextLinkDisabled}
                    onClick={this.lastPage}>
                        {'>>'}
                </span>
                {showStatus && <span className={styles.paginationStatus}>{`${activePage} of ${totalPages}`}</span>}
            </div>
        );
    }

    render() {
        return (
            <div>
                {this.generatePageLinks()}
            </div>
        );
    }
};

Pagination.PropTypes = {
    noOfItems: PropTypes.number.isRequired,
    noOfItemsPerPage: PropTypes.number,
    noOfPageLinksToDisplay: PropTypes.number,
    defaultActivePage: PropTypes.number,
    onPageChange: PropTypes.func.isRequired,
    showStatus: PropTypes.bool
};

Pagination.defaultProps = {
    noOfItemsPerPage: 10,
    noOfPageLinksToDisplay: 5,
    defaultActivePage: 1,
    showStatus: true
}

export default Pagination;