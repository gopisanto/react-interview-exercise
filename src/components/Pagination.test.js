import React from 'react';
import renderer from 'react-test-renderer';
import {configure, shallow, mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-15';
import sinon from 'sinon';
import Pagination from './Pagination';
import styles from './Pagination.css';

configure({ adapter: new Adapter() });

describe('<Pagination /> using snapshot', () => {
    it('renders nothing when no of pages is 1', () => {
        const tree = renderer.create(<Pagination noOfItems={10} />).toJSON();
    
        expect(tree).toMatchSnapshot();
    });
    it('renders with page links when more than one page', () => {
        const tree = renderer.create(<Pagination noOfItems={30} />).toJSON();
    
        expect(tree).toMatchSnapshot();
    });
});
describe('<Pagination /> using enzyme and sinon', () => {
    const noOfItems = 30;
    const noOfItemsPerPage = 3;

    it('sits on page 1 when defaultActivePage is not passed', () => {
        const wrapper = shallow(<Pagination noOfItems={noOfItems} />);
        const state = wrapper.state().activePage;
        expect(state).toEqual(1);
    });
    it('sits on page no which is passed as defaultActivePage', () => {
        const wrapper = shallow(<Pagination noOfItems={noOfItems} defaultActivePage={3} />);
        const state = wrapper.state().activePage;
        expect(state).toEqual(3);
    });
    it('active page is disabled for selection', () => {
        const wrapper = shallow(<Pagination noOfItems={noOfItems} defaultActivePage={2} />);
        const state = wrapper.state().activePage;
        expect(state).toEqual(2);
        expect(wrapper.find('span').findWhere(node => node.hasClass(styles.disabled)).length).toEqual(1);
    });
    it('active page, next (>), last (>>) are disabled for selection when activePage is last page', () => {
        const wrapper = shallow(
            <Pagination
                noOfItems={noOfItems}
                noOfItemsPerPage={noOfItemsPerPage}
                onPageChange={() => null}
            />
        );
        wrapper.instance().lastPage();
        expect(wrapper.find('span').findWhere(node => node.hasClass(styles.disabled)).length).toEqual(3);
    });
    it('last page is selected when >> is clicked', () => {
        const wrapper = shallow(
            <Pagination
                noOfItems={noOfItems}
                noOfItemsPerPage={noOfItemsPerPage}
                onPageChange={() => null}
            />
        );
        wrapper.instance().lastPage();
        expect(wrapper.state().activePage).toEqual(Math.ceil(30 / 3));
    });
    it('first page is selected when << is clicked', () => {        
        const wrapper = shallow(
            <Pagination
                noOfItems={noOfItems}
                noOfItemsPerPage={noOfItemsPerPage}
                onPageChange={() => null}
                defaultActivePage={4}
            />
        );
        wrapper.instance().firstPage();
        expect(wrapper.state().activePage).toEqual(1);
    });
    it('next page is selected when > is clicked', () => {        
        const wrapper = shallow(
            <Pagination
                noOfItems={noOfItems}
                noOfItemsPerPage={noOfItemsPerPage}
                onPageChange={() => null}
                defaultActivePage={2}
            />
        );
        wrapper.instance().nextPage();
        expect(wrapper.state().activePage).toEqual(3);
    });
    it('prev page is selected when < is clicked', () => {        
        const wrapper = shallow(
            <Pagination
                noOfItems={noOfItems}
                noOfItemsPerPage={noOfItemsPerPage}
                onPageChange={() => null}
                defaultActivePage={2}
            />
        );
        wrapper.instance().prevPage();
        expect(wrapper.state().activePage).toEqual(1);
    });
    it('when a page is selected, it should become the activePage', () => {        
        const wrapper = shallow(
            <Pagination
                noOfItems={noOfItems}
                noOfItemsPerPage={noOfItemsPerPage}
                onPageChange={() => null}
                defaultActivePage={2}
            />
        );
        wrapper.instance().handlePageClick(3)();
        expect(wrapper.state().activePage).toEqual(3);
    });
    it('calls onPageChange when page is selected', () => {
        const spyFunc = sinon.spy();
        const wrapper = mount(
            <Pagination
                noOfItems={noOfItems}
                noOfItemsPerPage={noOfItemsPerPage}
                onPageChange={spyFunc}
            />
        );
        wrapper.instance().lastPage();
        expect(spyFunc.callCount).toEqual(1);
    });
});