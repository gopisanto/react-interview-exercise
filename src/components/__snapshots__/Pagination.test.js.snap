exports[`<Pagination /> using snapshot renders nothing when no of pages is 1 1`] = `<div />`;

exports[`<Pagination /> using snapshot renders with page links when more than one page 1`] = `
<div>
  <div
    className={undefined}>
    <span
      className="undefined"
      disabled={true}
      onClick={[Function]}>
      <<
    </span>
    <span
      className="undefined"
      disabled={true}
      onClick={[Function]}>
      <
    </span>
    <span
      className="undefined"
      onClick={[Function]}>
      1
    </span>
    <span
      className=""
      onClick={[Function]}>
      2
    </span>
    <span
      className=""
      onClick={[Function]}>
      3
    </span>
    <span
      className=""
      disabled={false}
      onClick={[Function]}>
      >
    </span>
    <span
      className=""
      disabled={false}
      onClick={[Function]}>
      >>
    </span>
    <span
      className={undefined}>
      1 of 3
    </span>
  </div>
</div>
`;
